Re-enable Sentry on Android, with `enableNdk: false`

chrisbobbe opened the pull request:
~~Our hope here is to fix the Android crashes in #5757. Details in the commit message.~~

~~Edit, following https://github.com/zulip/zulip-mobile/pull/5765#issuecomment-1728612489: This did not fix it. [Subsequent testing](https://chat.zulip.org/#narrow/stream/48-mobile/topic/Android.20crashes/near/1644202) showed that the crash still happens at this later version.~~

Edit, following https://github.com/zulip/zulip-mobile/pull/5765#issuecomment-1745954363: This PR was originally just an upgrade of `@sentry/react-native`. It still does that, but goes to a later version, where they added an `enableNdk` option, and it uses that option for the sake of #5766.

Fixes: #5766

chrisbobbe commented:
Closing, as our initial testing was found to be invalid. The DSN was accidentally not included, so the app wasn't told where to send Sentry events. In a subsequent test in which the DSN was included, the crash reproduced: https://chat.zulip.org/#narrow/stream/48-mobile/topic/Android.20crashes/near/1644202

chrisbobbe commented:
Ah, as Greg points out in [chat](https://chat.zulip.org/#narrow/stream/48-mobile/topic/Android.20crashes/near/1644205):
> We'll still want the upgrade #M5765 — if Sentry comes out with a fix, we'll be best placed to take it having already upgraded to the current latest. But we don't need to rush to merge it today.

So I'll reopen, but I'll update the branch and PR description to not say it fixes the crash.

gnprice commented:
OK, I spent some time today investigating this. I've just pushed back to this PR a branch with several further changes:
* Upgrade to the newer 5.10.0, which adds an option `enableNdk`
* Set that option to false on Android
* Revert e8c57e799ba03f09e75114ff74a3d4e61594a358 / #5768, to re-enable Sentry for our normal release builds on Android, fixing #5766

That's based on some testing which confirmed that `enableNdk: false` appears to resolve the crashes, and that we do get Sentry events after these changes. Details in chat thread: https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/re-enabling.20Sentry/near/1654107

---

Then for the types, I spent a bit of time trying to get working type definitions through TsFlower. My conclusion is that given that we're planning to mothball this codebase soon, it's not worth it. Instead, we just keep the existing hand-edited type definitions we have and add the one new piece of API we use (namely that `enableSdk` option).

In particular, there's a fairly small API surface area that we actually use from Sentry; and it's pretty infrequent (was pretty infrequent, even when zulip-mobile was the codebase we were developing full time) that we make changes in the parts of our codebase that interact with Sentry directly. So the value from having completely accurate type definitions for Sentry is fairly low.

And conversely, the internals of the Sentry type definitions are fairly complex, type-system-wise. It's possible that if we buckled down to it, then adding to TsFlower whatever it needs in order to handle them wouldn't be too hard… but I'm pretty sure it'd be a few hours, at minimum. And given the above, that's more work than it's worth to us.

chrisbobbe commented:
wip/ref Revert "sentry: Disable Sentry for now on Android"

What's left to do here, indicated by "wip/ref"? 🙂

chrisbobbe commented:
Thanks for that work! Pushed back some changes including:

- Removed some "wip"s in commit messages
- Added back the commit to have us upload Android debug symbols

gnprice commented on android/app/build.gradle:
```diff
@@ -85,8 +85,26 @@ project.ext.react = [
     // Conditionally enable Sentry, because it requires an auth key.
     apply from: "../../node_modules/@sentry/react-native/sentry.gradle"
+
+    apply plugin: "io.sentry.android.gradle"
+    // https://docs.sentry.io/platforms/android/configuration/gradle/#configure
+    sentry {
+        uploadNativeSymbols = true
+        includeNativeSources = true
```
I think "native" here refers to actual native code, i.e. machine code, the kind whose source code might be C or C++ or Rust.

(In a React Native context on Android the term "native" is ambiguous, as it often means JVM code like one writes in Java or Kotlin. But here this is Sentry's Gradle plugin for Android in general, not specific to React Native; and in a general Android context, JVM code is just the normal kind of code, and "native" always means machine code.)

In that case, these symbols and sources are things that could be useful for interpreting events reported for exceptions in native code on Android — but reporting those events is exactly the functionality we turn off in our Sentry config later in this PR.

So I think this piece of configuration won't actually do anything for us, and we can leave out this commit.

(If we were including it, I'd want to spend a bit more time looking through what this Gradle plugin does and validating that it seems unlikely to break anything. For example, one of the listed features at https://docs.sentry.io/platforms/android/configuration/gradle/ is "Auto-instrumentation tracing through bytecode manipulation", which sounds like something that could cause hard-to-debug problems if buggy; and from the config docs lower down on that page, it looks like that's enabled by default.)

gnprice commented on src/sentry.js:
```diff
@@ -153,10 +154,22 @@ export const initializeSentry = () => {
+            // memory-corruption bugs that cause the app to crash.
+            //   https://github.com/zulip/zulip-mobile/issues/5766
+            //   https://github.com/getsentry/sentry-java/issues/2955#issuecomment-1739030872
+            enableNdk: false,
```
Greg found in testing that this, plus re-enabling Sentry on Android
    (coming up) seems adequate for #5766.

In this commit message let's include a link for that testing, since that's easy to do:
https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/re-enabling.20Sentry/near/1654140

gnprice reviewed (commented):
Thanks! Two comments below.

I'll go ahead and merge with these changes, unless you think the "Turn on debug-symbol uploads" commit is essential and there's something else that should be adjusted if we leave it out.

chrisbobbe commented:
> I'll go ahead and merge with these changes, unless you think the "Turn on debug-symbol uploads" commit is essential and there's something else that should be adjusted if we leave it out.

Sounds good, please go ahead and merge with those changes! 🙂 Thanks.
